File: backend/src/services/modules.service.ts
import { ModulesRepository, CourseModule } from '../repositories/modules.repository';

export class ModulesService {
  private modulesRepo: ModulesRepository;

  constructor() {
    this.modulesRepo = new ModulesRepository();
  }

  /**
   * Obtener módulos de un curso
   */
  async getModulesByCourse(courseId: string): Promise<CourseModule[]> {
    return this.modulesRepo.findByCourse(courseId);
  }

  /**
   * Obtener un módulo por ID
   */
  async getModule(moduleId: string): Promise<CourseModule | null> {
    return this.modulesRepo.findById(moduleId);
  }

  /**
   * Crear módulo para un curso
   */
  async createModule(courseId: string, module: Omit<CourseModule, 'id' | 'course_id' | 'created_at' | 'updated_at'>): Promise<string> {
    return this.modulesRepo.createForCourse(courseId, module);
  }

  /**
   * Actualizar módulo
   */
  async updateModule(moduleId: string, updates: Partial<CourseModule>): Promise<void> {
    await this.modulesRepo.update(moduleId, updates);
  }

  /**
   * Eliminar módulo
   */
  async deleteModule(moduleId: string): Promise<void> {
    await this.modulesRepo.delete(moduleId);
  }
}

File: backend/app/services/modules_service.py
"""Modules service."""

from app.repositories.modules_repository import ModulesRepository


class ModulesService:
    def __init__(self, repository: ModulesRepository | None = None) -> None:
        self._repository = repository or ModulesRepository()

    def list_modules(self, course_id: str) -> list[dict]:
        return self._repository.list_by_course(course_id)

    def get_module(self, module_id: str) -> dict | None:
        return self._repository.get(module_id)

    def create_module(self, course_id: str, module_data: dict) -> str:
        # Ensure course_id link
        module_data["course_id"] = course_id
        return self._repository.create(module_data)

    def update_module(self, module_id: str, updates: dict) -> None:
        return self._repository.update(module_id, updates)

    def delete_module(self, module_id: str) -> None:
        return self._repository.delete(module_id)
